clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -1124,6 +1126,62 @@ public void testFetchMaxPollRecords() {
         assertEquals(5, recordsToTest.get(1).offset());
     }
 
+    /**
+     * KAFKA-15836:
+     * Test that max.poll.records is honoured when consuming from multiple topic-partitions and the
+     * fetched records are not aligned on max.poll.records boundaries.
+     *
+     * tp0 has records 1,2,3; tp1 has records 6,7,8
+     * max.poll.records is 2
+     * 
+     * poll 1 should return 1,2
+     * poll 2 should return 3,6
+     * poll 3 should return 7,8
+     * 
+     * Or similar :)
```

Comment:
I'm pretty sure that there's nothing in the current code that will attempt to pull _N_ records from _M_ partitions fairly. Is it possible that the number of records produced to topic _T1_ is >= the value of `maxRecords` at any given point in time, records produced to _T2_ will never be consumed?

Reply:
Well, both client and server do have some kind of next-in-line behaviour I think. Not guaranteed, but it does at least attempt to avoid starvation.